### PR TITLE
Rewrite battleship processor in Rust

### PR DESCRIPTION
**Pull request by knkski**
Replaces the Python Battleship TP/CLI with a Rust version.

Signed-off-by: Kenneth Koski <knkski@bitwise.io>

### CODE REVIEW

File: families/battleship/tests/test_tp_battleship.yaml [hunk 1]
```diff
@@ -1,61 +0,0 @@
...
-    image: sawtooth-battleship-tp$INSTALL_TYPE:$ISOLATION_ID
-    volumes:
-      - $SAWTOOTH_CORE:/project/sawtooth-core
-    expose:
-      - 4004
-    command: battleship-tp-python -vv -C tcp://test-tp-battleship:4004
-    stop_signal: SIGKILL
-
-  test-tp-battleship:
-    build:
```
[the remark this answers is not in the document]
Comment (knkski) [reply families/battleship/tests/test_tp_battleship.yaml]: @agunde406: This test relies on Python files that are getting deleted in this PR. Are you saying we should keep those files?

[the remark this answers is not in the document]
Comment (agunde406) [reply families/battleship/tests/test_tp_battleship.yaml]: That is what has been done in for the other tp that have been rewritten so far.

File: families/battleship/Dockerfile [hunk 2]
```diff
@@ -12,38 +12,36 @@
...
+RUN curl https://sh.rustup.rs -sSf > /usr/bin/rustup-init \
+ && chmod +x /usr/bin/rustup-init \
+ && rustup-init -y
 
 WORKDIR /project/sawtooth-core
+
+CMD cd families/battleship \
+ && echo "\033[0;32m--- Building battleship ---\n\033[0m" \
+ && cargo build --release \
+ && cp ./target/release/processor /project/sawtooth-core/bin/battleship-tp \
```
Comment (boydjohnson) [inline families/battleship/Dockerfile]: I think `./target/release/processor` should be `./target/release/battleship`.

Comment (knkski) [reply families/battleship/Dockerfile]: I've got the binaries under `src/bin/` as `processor` and `cli`, so unless we change those, I think this needs to be `./target/release/processor`.

File: families/battleship/tests/test_tp_battleship.yaml [hunk 3]
```diff
@@ -44,16 +53,18 @@ services:
...
       - 4004
-    command: nose2-3
-        -c /project/sawtooth-core/families/battleship/nose2.cfg
-        -v
-        -s /project/sawtooth-core/families/battleship/tests
-        test_tp_battleship
     stop_signal: SIGKILL
     environment:
-        TEST_BIND: "tcp://eth0:4004"
+        TEST_BIND: "tcp://0.0.0.0:4004"
```
Comment (boydjohnson) [inline families/battleship/tests/test_tp_battleship.yaml]: Do we know why this change is necessary? It is generally bad form to bind to all interfaces.

File: families/battleship/src/game.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+/// Get a state address for a given game of battleship
+pub fn get_battleship_address(name: &str) -> String {
+    vec![get_battleship_prefix(), get_battleship_name_hash(name)].join("")
+}
+
+/// Hash the value and nonce for a revealed space
+/// When a player joins, we're given a hashed board of their ship locations.
+/// On each player's turn, they give us a nonce and the value of that space,
+/// e.g. `'-'` and `"g54h45hwer"`. We hash those together and compare against
```
Comment (boydjohnson) [inline families/battleship/src/game.rs]: Nice explanation.

File: families/battleship/tests/test_tp_battleship.yaml (hunk 3 above)
Comment (knkski) [reply families/battleship/tests/test_tp_battleship.yaml]: @boydjohnson: Not sure why I had that there, removed that change

File: bin/run_tests [hunk 5]
```diff
@@ -202,6 +202,7 @@ test_block_info() {
 
 test_battleship() {
   run_docker_test ./families/battleship/battleship_unit_tests.yaml
+  run_docker_test ./families/battleship/tests/test_tp_battleship.yaml
```
Comment (agunde406) [inline bin/run_tests]: What's the reasoning for removing them and then adding them back? Instead of just not removing them?

Comment (knkski) [reply bin/run_tests]: It seems weird to me to have a commit that has tests that won't run due to missing code. Given that and the fact that there were changes to the tests, it felt easier to just re-introduce them and make those changes in a separate commit.

Comment (agunde406) [reply bin/run_tests]: I agree that it is weird that the commit has the test that will fail but it also has two downsides.  It removes the history from the original tests and makes it hard to see what the changes you made to the tests are. I would suggest not removing the tests to begin with.

File: bin/run_tests [hunk 6]
```diff
@@ -196,6 +196,10 @@ test_block_info() {
   copy_coverage .coverage.block_info_injector
 }
 
+test_battleship() {
```
Comment (agunde406) [inline bin/run_tests]: Please add a description to the commit to explain the changes

File: families/battleship/tests/battleship_board.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
+# Copyright 2016 Intel Corporation
```
Comment (agunde406) [inline families/battleship/tests/battleship_board.py]: This should be bitwise 2018 if this is a new file

File: families/battleship/tests/battleship_message_factory.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Copyright 2017 Intel Corporation
```
Comment (agunde406) [inline families/battleship/tests/battleship_message_factory.py]: This should be bitwise 2018 if this is a new file

File: families/battleship/tests/test_tp_battleship.py [hunk 9]
```diff
@@ -15,21 +15,30 @@
 
 import logging
 import json
-
 from sawtooth_sdk.protobuf.validator_pb2 import Message
-from sawtooth_processor_test.transaction_processor_test_case \
```
Comment (agunde406) [inline families/battleship/tests/test_tp_battleship.py]: Since this is significant changes, you should add a bitwise copyright underneath the intel on

**Comment by knkski** [comment]
@agunde406: Updated with those changes
